frontend/build.rs

```diff
@@ -0,0 +1,107 @@
+///
+/// Build Script
+/// This is run as a pre-build step -- before the rust backend is compiled.
+/// NOTE: Should be included in root's build script
+///
+use std::ffi::OsStr;
+use std::fmt::Display;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+use std::{io::Write, process::Command};
+
+#[cfg(target_family = "windows")]
+const SHELL: &str = "cmd";
+#[cfg(target_family = "unix")]
+const SHELL: &str = "sh";
+#[cfg(target_family = "windows")]
+const SHELL_ARG: &str = "/C";
+#[cfg(target_family = "unix")]
+const SHELL_ARG: &str = "-c";
+
+/*
+ * Note 1: this file was written for *nix systems -- it likely won't
+ * work on windows!
+ */
+
+#[allow(dead_code)]
+fn shell(command: impl AsRef<OsStr> + Display) {
+    // println!("build.rs => {}", command);
+
+    let output = Command::new(SHELL)
+        .arg(SHELL_ARG)
+        .arg(&command)
+        .output()
+        .expect(format!("Failed to run {cmd}", cmd = command).as_str());
+
+    // println!("build.rs => {:?}", output.stdout);
+    let mut file = File::create("build.log").expect("Couldn't create file...");
+    file.write(b"build log\n\n\n\nSTDOUT:\n")
+        .expect("Couldn't write to build log");
+    file.write_all(&output.stdout)
+        .expect("Couldn't write to build log");
+    file.write(b"\n\n\n\nSTDERR:\n")
+        .expect("Couldn't write to build log");
+    file.write_all(&output.stderr)
+        .expect("Couldn't write to build log");
+}
+
+pub fn build_types() {
+    let mut inputs = vec![PathBuf::from(env!("CARGO_MANIFEST_DIR"))];
+    let mut output = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    inputs[0].pop();
+
+    inputs[0].push("backend");
+    output.push("src/types/rust.d.ts");
+
+    tsync::generate_typescript_defs(inputs, output, false);
+}
+
+pub fn build_frontend() {
+    // Only install frontend dependencies when building release
+    // #[cfg(not(debug_assertions))]
+    shell("yarn");
+
+    // Only build frontend when building a release
+    // #[cfg(not(debug_assertions))]
+    shell("yarn build");
+
+    let mut dir = PathBuf::from(std::env::var("OUT_DIR").unwrap()); // OUT_DIR == <project_dir>/target/<target_profile>/build/bob_management-<HASH>/out
+    dir.pop();
+    dir.pop();
+    dir.pop();
+    dir.pop();
+    let mut project_dir = dir.clone();
+    let mut target = dir.clone();
+    project_dir.pop();
+    target.push("dist");
+    println!("cargo:warning=Moving /dist to: {dir:?}");
+    println!("cargo:warning=PROJECT DIR: {project_dir:?}");
```

Comment:
These 2 `println!` are informational (they are not a warnings). If we will decide at some point to treat warnings as errors this can become a problem

Reply:
There are no other way to inform user where output files were moved (in terminal). Print that info in `build.log`, perhaps?

Will be changed

Reply:
You can use `println!` without `cargo:warning=`

Reply:
[not really](https://github.com/rust-lang/cargo/issues/985#issuecomment-1071667472)

For some reason they decided, that output in build scripts is too "noisy"

Reply:
Ok. I see, that you write information into `build.log`. I think, that's ok too